Examples/Io/HepMC3/src/HepMC3Writer.cpp

```diff
@@ -111,22 +128,111 @@ ProcessCode HepMC3Writer::writeT(
 
     ACTS_VERBOSE("Writing per-event file " << perEventFile);
     auto writer = createWriter(perEventFile);
-    auto result = write(*writer);
+
+    writer->write_event(*event);
+    auto result = ProcessCode::SUCCESS;
+    if (writer->failed()) {
+      ACTS_ERROR("Failed to write event number: " << ctx.eventNumber);
+      result = ProcessCode::ABORT;
+    }
     writer->close();
     return result;
+  }
+
+  if (!m_cfg.writeEventsInOrder) {
+    std::scoped_lock lock{m_mutex};
+    // Unconditionally write events in whatever order they come in
+    m_writer->write_event(*event);
+    if (m_writer->failed()) {
+      ACTS_ERROR("Failed to write event number: " << ctx.eventNumber);
+      return ProcessCode::ABORT;
+    }
+    return ProcessCode::SUCCESS;
+  }
+
+  std::scoped_lock lock{m_mutex};
+
+  auto printQueue = [&]() {
+    ACTS_VERBOSE("queue=[" << [&]() {
+      std::vector<std::string> numbers;
+      numbers.reserve(m_eventQueue.size());
+      std::ranges::transform(
+          m_eventQueue, std::back_inserter(numbers),
+          [](const auto& pair) { return std::to_string(pair.first); });
+
+      return boost::algorithm::join(numbers, ", ");
+    }() << "]");
+  };
+
+  if (ctx.eventNumber == m_nextEvent) {
+    ACTS_DEBUG("event_nr=" << ctx.eventNumber
+                           << " is the next event -> writing");
+
+    // write
+    m_writer->write_event(*event);
+    if (m_writer->failed()) {
+      ACTS_ERROR("Failed to write event number: " << ctx.eventNumber);
+      return ProcessCode::ABORT;
+    }
+    m_nextEvent++;
+
+    std::size_t nWritten = 1;
+
+    printQueue();
+
+    while (!m_eventQueue.empty() &&
+           m_eventQueue.front().first == static_cast<long long>(m_nextEvent)) {
+      auto [nextEventNumber, nextEvent] = std::move(m_eventQueue.front());
+      ACTS_VERBOSE("Writing event number: " << nextEventNumber);
+      m_eventQueue.erase(m_eventQueue.begin());
+
+      m_writer->write_event(*nextEvent);
+      if (m_writer->failed()) {
+        ACTS_ERROR("Failed to write event number: " << nextEventNumber);
+        return ProcessCode::ABORT;
```

Comment:
Question: Does returning `ABORT` here end the entire program? If not, you may want to release the semaphore here.

Reply:
It won't abort right away, but the scheduler will throw from the body of the parallel for, and that will terminate the job. So since the `ABORT` is not recoverable, I don't think we need to relase the semaphore.